include/pybind11/pybind11.h

```diff
@@ -1378,8 +1378,10 @@ iterator make_iterator(Iterator first, Sentinel last, Extra &&... extra) {
                     ++s.it;
                 else
                     s.first = false;
-                if (s.it == s.end)
+                if (s.it == s.end) {
+                    s.first = true;
```

Comment:
It was a bit non-obvious to me at first glance why this works, which I think may have been due to the `first` name; a rename of `first` to something like `first_or_done` would be more self-documenting.